kernel/src/grant.rs

```diff
@@ -129,9 +130,207 @@ use core::slice;
+
+        let allow_rw_num = allow_ro_array.add(allow_ro_num.read()) as *mut usize;
+        let allow_rw_array = allow_rw_num.add(1) as *mut SavedAllowRw;
+
+        Self {
+            upcalls_num,
+            upcalls_array,
+            allow_ro_num,
+            allow_ro_array,
+            allow_rw_num,
+            allow_rw_array,
+        }
+    }
+
+    /// Creates a layout from the specified pointer and lengths of arrays
+    ///
+    /// # Safety
+    ///
+    /// The incoming base pointer must be well aligned but does not need to point
+    /// to initialized data
+    unsafe fn from_counts(
+        base_ptr: *mut u8,
+        upcalls_num_val: UpcallItems,
+        allow_ro_num_val: AllowRoItems,
+    ) -> Self {
+        let upcalls_num = base_ptr as *mut usize;
+        let upcalls_array = upcalls_num.add(1) as *mut SavedUpcall;
+
+        let allow_ro_num = upcalls_array.add(upcalls_num_val.0) as *mut usize;
+        let allow_ro_array = allow_ro_num.add(1) as *mut SavedAllowRo;
+
+        let allow_rw_num = allow_ro_array.add(allow_ro_num_val.0) as *mut usize;
+        let allow_rw_array = allow_rw_num.add(1) as *mut SavedAllowRw;
+
+        Self {
+            upcalls_num,
+            upcalls_array,
+            allow_ro_num,
+            allow_ro_array,
+            allow_rw_num,
+            allow_rw_array,
+        }
+    }
+
+    /// Returns the entire grant size including the kernel own memory, padding, and data for T.
+    /// Requires that grant_t_align be a power of 2, which is guaranteed from align_of rust calls
+    fn grant_size(
+        upcalls_num: UpcallItems,
+        allow_ro_num: AllowRoItems,
+        allow_rw_num: AllowRwItems,
+        grant_t_size: GrantDataSize,
+        grant_t_align: GrantDataAlign,
+    ) -> usize {
+        let kernel_managed_size = 3 * size_of::<usize>()
+            + upcalls_num.0 * size_of::<SavedUpcall>()
+            + allow_ro_num.0 * size_of::<SavedAllowRo>()
+            + allow_rw_num.0 * size_of::<SavedAllowRw>();
+        // We know that grant_t_align is a power of 2, so we can make a mask that will save
+        // only the remainder bits
+        let grant_t_align_mask = grant_t_align.0 - 1;
+        // Determine padding to get to the next multipe of grant_t_align by taking the remainder
+        // and subtracting that from the alignment, then ensuring a full alignment value maps to 0
+        let padding =
+            (grant_t_align.0 - (kernel_managed_size & grant_t_align_mask)) & grant_t_align_mask;
+        kernel_managed_size + padding + grant_t_size.0
+    }
+
+    /// Returns the alignment of the entire grant region based on the alignment of data T
+    fn grant_align(grant_t_align: GrantDataAlign) -> usize {
+        // The kernel owned memory all aligned to usize. We need to use the higher of the two
+        // alignment to ensure our padding calculations work for any alignment of T
+        cmp::max(size_of::<usize>(), grant_t_align.0)
```

Comment:
Should this be an `align_of::<usize>` rather than `size_of::<usize>`?

Reply:
Good call; updated to `align_of`.